internal/nexodus/proxyrule.go

```diff
@@ -0,0 +1,138 @@
+package nexodus
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+)
+
+type ProxyType int
+
+const (
+	ProxyTypeEgress ProxyType = iota
+	ProxyTypeIngress
+)
+
+func (ruleType ProxyType) String() string {
+	switch ruleType {
+	case ProxyTypeEgress:
+		return "egress"
+	case ProxyTypeIngress:
+		return "ingress"
+	default:
+		panic(fmt.Sprintf("Invalid proxy rule type: %d", ruleType))
+	}
+}
+
+type ProxyProtocol string
+
+const (
+	proxyProtocolTCP ProxyProtocol = "tcp"
+	proxyProtocolUDP ProxyProtocol = "udp"
+)
+
+func parseProxyProtocol(protocol string) (ProxyProtocol, error) {
+	switch strings.ToLower(protocol) {
+	case "tcp":
+		return proxyProtocolTCP, nil
+	case "udp":
+		return proxyProtocolUDP, nil
+	default:
+		return "", fmt.Errorf("Invalid protocol (%s)", protocol)
+	}
+}
+
+type ProxyKey struct {
+	ruleType   ProxyType
+	protocol   ProxyProtocol
+	listenPort int
+}
+
+func (rule ProxyKey) String() string {
+	return fmt.Sprintf("%s:%s:%d", rule.ruleType, rule.protocol, rule.listenPort)
+}
+
+type ProxyRule struct {
+	ProxyKey
+	dest   HostPort
+	stored bool
+}
+
+type HostPort struct {
+	host string
+	port int
+}
+
+func (hp HostPort) String() string {
+	return fmt.Sprintf("%s:%d", hp.host, hp.port)
+}
+
+func (rule ProxyRule) String() string {
+	// protocol:port:destination_ip:destination_port
+	return fmt.Sprintf("%s:%d:%s:%d", rule.protocol, rule.listenPort, rule.dest.host, rule.dest.port)
+}
+
+func (rule ProxyRule) AsFlag() string {
+	return fmt.Sprintf("--%s %s", rule.ruleType, rule)
+}
+
+func parsePort(portStr string) (int, error) {
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return 0, fmt.Errorf("Invalid port (%s): %w", portStr, err)
+	}
+	return port, nil
+}
+
+func ParseProxyRule(rule string, ruleType ProxyType) (emptyRule ProxyRule, err error) {
+	// protocol:port:destination_ip:destination_port
+	parts := strings.Split(rule, ":")
+	if len(parts) < 4 {
+		return emptyRule, fmt.Errorf("Invalid proxy rule format, must specify 4 colon-separated values (%s)", rule)
+	}
+
+	protocol, err := parseProxyProtocol(parts[0])
+	if err != nil {
+		return emptyRule, err
+	}
+
+	port, err := parsePort(parts[1])
+	if err != nil {
+		return emptyRule, err
+	}
+	if port < 1 || port > 65535 {
+		return emptyRule, fmt.Errorf("Invalid port (%d)", port)
+	}
```

Comment:
This is probably old code that you just moved, but this should probably move into `parsePort()` if you don't mind while you're in here anyway.

Reply:
FYI: messages for `error` types should always start lower case since you often prepend more context  to them in wrapper errors.  So I'll change that too.

Reply:
ah, didn't know, thanks makes sense - thanks!